### DISCUSSION TITLE
Making non-fullCube blocks opaque

### DISCUSSION THREAD

**Discussion by sprucerooth** [FabricMC/fabric-api#1767] [Mod Dev Support]
I've been scratching my head and debugging this for a few days now.

I have a custom block with a less than full cuboid outline shape (a vertical slab). How can I make it be opaque? 

As soon as I override AbstractBlock.getOutlineShape to return anything else than a VoxelShapes.fullCube, the block lets light pass through it.

**Comment by Technici4n** [comment]
If I recall correctly, you can override a method in your block, like `getOpacity`. There may also be something in the block settings.

**Reply by sprucerooth** [reply]
Thank you ! That was it. Silly that I missed that one.

One issue still remains with the lighting.

This is a wall of vertical slabs with the last piece missing:
![smooth-light](https://user-images.githubusercontent.com/7568776/136710702-aa13f224-471a-41af-b089-48371d392c6d.png)

When I place the last slab, the room now turns dark as it should. However, there is a slight light bleed from outside on the remaining half of the adjacent blocks (the slab takes up one half).
![smooth-light-whole](https://user-images.githubusercontent.com/7568776/136710752-1401a184-e844-46c4-96e6-a9363039119b.png)

Also, if I turn off smooth lighting, this happens:
![smooth-light-off](https://user-images.githubusercontent.com/7568776/136710770-02b00759-11b3-4c04-b9bb-d6434e61ce7d.png)
The part of the adjacent blocks that are dark here is also the area where the light bleed happens in the second image.

So what kind of light related fix might I be missing?

**Comment by warjort** [comment]
I am far from an expert on this, but have you got something like the following for your block?
````
   @Override
   public boolean hasSidedTransparency(BlockState state) {
      return true;
   }
````
Most of the vanilla blocks that are not full cubes change this property. e.g. the vanilla slab does:
````
   public boolean hasSidedTransparency(BlockState state) {
      return state.get(TYPE) != SlabType.DOUBLE;
   }
````
i.e. it is false when the double slab occupies the full block, true for half slabs.

**Reply by sprucerooth** [reply]
I noticed that as well and I tried it but no luck :/ I'm not really sure what sided transparency means.

**Reply by sprucerooth** [reply]
I could find a similar problem in Forge on 1.12:
https://forums.minecraftforge.net/topic/63248-112-half-slabs-display-issue-when-smooth-lighting-is-off/

Might be completely irrelevant since it's Forge and an older version.

**Reply by warjort** [reply]
> I noticed that as well and I tried it but no luck :/ I'm not really sure what sided transparency means.

I believe it is used to do extra checks for lighting updates. I suggested it because other non full cube blocks change it.

My guess would be that you have some mismatch between the block model and getOutlineShape(), e.g. the coordinates of your shape or the culled faces. But like I said above, I am no expert on rendering.

Maybe you can post the block model and definition for your getOutlineShape()? More eyes might see an error if there is one.

I would also suggest to compare your code against vanilla's SlabBlock and its block models. Your block sounds similar to it?

**Reply by sprucerooth** [reply]
Well the model and the outlines match up. Of course I could be missing something but I don't think that's the issue.
Yeah I have studied SlabBlock before as well but I can't find any code that seems to be related to that.

I tested the same thing but using tinted glass, which is supposed to block light.
Here I have disabled smooth lighting. The dark areas are there as they were with my slabs(though covered with the transparent glass block):
![tinted-smooth-off](https://user-images.githubusercontent.com/7568776/136761366-de8c1b97-8734-449b-8f69-48a49777c886.png)

And if I enable smooth lighting:

![tinted](https://user-images.githubusercontent.com/7568776/136761568-2cb59798-86a7-4cf8-802c-11a73d54e446.png)
As you can see that light gradient/bleed is there on the adjacent blocks, just like with my vertical slabs. Only now the adjacent blocks are completely covered by the tinted glass block since its a full cube.

So the behaviour seems intended.
However this does not happen on regular slabs so there has to be a way to disable it!
